service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectory.java

```diff
@@ -0,0 +1,106 @@
+package uk.nhs.adaptors.gp2gp.ehr.mapper;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.hl7.fhir.dstu3.model.Bundle;
+import org.hl7.fhir.dstu3.model.PractitionerRole;
+import org.hl7.fhir.dstu3.model.Reference;
+import org.hl7.fhir.dstu3.model.ResourceType;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import uk.nhs.adaptors.gp2gp.common.service.RandomIdGeneratorService;
+import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
+
+@Slf4j
+@AllArgsConstructor
+public class AgentDirectory {
+    private final RandomIdGeneratorService randomIdGeneratorService;
+    private final Map<AgentKey, String> agentDirectory = new HashMap<>();
+    private final Bundle inputBundle;
+
+    public Set<Map.Entry<AgentKey, String>> getEntries() {
+        return agentDirectory.entrySet();
+    }
+
+    public String getAgentId(Reference reference) {
+        if (reference.hasReferenceElement()) {
+            String resourceType = reference.getReferenceElement().getResourceType();
+            if (ResourceType.Practitioner.name().equals(resourceType)) {
+                return putIfAbsent(buildAgentKeyForPractitioner(reference));
+            } else if (ResourceType.Organization.name().equals(resourceType)) {
+                return putIfAbsent(AgentKey.builder().organizationReference(reference.getReference()).build());
+            }
+        }
+
+        LOGGER.error("Invalid agent reference to not supported resource: " + reference.getReference());
```

Comment:
Don't need to log and throw, can just throw with an appropriate message and let a downstream error handler log it out.